pAI's have headsets and can use encryption keys

RaveRadbury opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
From the holy page `/code/modules/mob/living/silicon/pai/software.dm`
`// TODO:
//	- Additional radio modules
//	- Potentially roll HUDs and Records into one
//	- Shock collar/lock system for prisoner pAIs?
//  - Put cable in user's hand instead of on the ground
//  - Camera jack`

This PR accomplishes one of the Precursor TODO's: Additional Radio Modules

## About The Pull Request

This gives pAI's access to speaking on departmental channels. Any encryption key you can add to a headset will do. They will need to download the program before it will work.

- pAI's spawn with an empty version of the AI headset 
  - It cannot use high-volume  
  - It acts like a station bounced radio when comms are down.
- You can add encryption up to 2 keys 
  - Add them to the pAI by using the key on the pAI card. 
  - Remove them by using the screwdriver on the pAI card. 
  - While in a bot the keys are unavailable to you, but access returns on eject
  - It will hold binary keys, but will not grant binary access
- It will cost 20 RAM to allow encryptionkeys to be added to the pAI (I welcome feedback on pricing)
 
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Currently these little guys are limited to common and eavesdropping headsets by occupying the same tile. If AI's get to have headsets, then surely pAI's can have that same tech. This PR adds more use to encryption keys and also creates a more interesting pAI experience, as they can be slotted with 2 keys for their own use.

- This enables pAI's to have better communication with other people in their master's department.
- You can match pAI's with med or sec huds and the corresponding keys. More synergies with existing abilities.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: pAI's now have the same headset tech as AI. Add and remove keys just like you would with headsets.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

ToDo
- [x] pAI's spawn with headset that can be used like a Station Bounced Radio (unique)
- [x] the paicard can be interacted with like a headset re: encryption keys
- [x] No high volume allowed
- [x] Slap a RAM price tag on it 
- [x] Integrate it into the software stack

RaveRadbury committed:
pAI's have little AI headsets and can use encryption keys.

ATH1909 commented:
What happens if a captain's headset's encryption key is inserted into a pAI? Can the pAI only use the security and command channels that are on by default, or can they turn on all of the other channels (that the captain's headset has access to) through an interface of some kind?

RaveRadbury commented:
I am new to BYOND coding so it's pretty barebones. If the Captain's Headset can eject it's encryption key with the additional channels being stored in the key then yeah, you might be able to squeeze out more than just Command and Sec. No promises.

I could likely try and set up a menu for the 3rd part of all of this, wherein I've reintegrated the radio and have a better idea of what I'm doing.

ATH1909 commented:
> I am new to BYOND coding so it's pretty barebones. If the Captain's Headset can eject it's encryption key with the additional channels being stored in the key then yeah, you might be able to squeeze out more than just Command and Sec. No promises.
> 
> I could likely try and set up a menu for the 3rd part of all of this, wherein I've reintegrated the radio and have a better idea of what I'm doing.

Borgs need to turn on those channels again (even if they were turned on before the key was removed from its previous headset) after having a captain's key installed (if they want to hear those extra channels), which is why I asked.

GunHog commented:
Just make sure that the pAI can access the radio's controls, and that will not be an issue. The pAI player could just choose its own active channels.

RaveRadbury commented:
@GunHog This is the 4th thing that silicon headsets do that is crucial to pAI headset functioning.
`/obj/item/radio/headset/silicon/can_receive(freq, level)
	return ..(freq, level, TRUE)
`

GunHog commented:
Oh, it is the check for AIuser. I saw that, did not remember what it did. My bad. Do set `command` to `FALSE`, though. Yelling ability has to be restricted.

RaveRadbury committed:
pAI's now spawn radio AND headset.

Reverted radio code and changed new variables to the new headset var

RaveRadbury committed:
Undid the comment out of radio code.

As far as I can tell this isn't working properly though. Tried testing it locally and can't enable/disable radio as pAI or master.

RaveRadbury commented:
Did the code proper and added a headset var rather than piggybacking on the radio. But now that I look at the radio stuff, I can't even disable radios as master and as pAI my radio settings are all grayed out ~but it's not really any of my business since I didn't write it.~

ExcessiveUseOfCobblestone commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -255,12 +255,17 @@ GLOBAL_LIST_INIT(channel_tokens, list(
+/obj/item/radio/headset/silicon/pai
+	name = "\proper personal Integrated Subspace Transceiver "
+	keyslot = null
+	keyslot2 = null
```
are these not the default values?

Also move the base procs above the variants (I don't even see a `/obj/item/radio/headset/silicon` definition?)

RaveRadbury replied:
Oh yeah, I guess those are redundancies. 4dplanner recommended that I set up the /silicon/ base so I went for it rather than trying to go headset/ai/pai.

RaveRadbury committed:
Removed redundant null keyslots.

Fox-McCloud commented:
Do we really need to buff personalized harm alarms to make them married even tighter to important high ranking channels like command and security?

RaveRadbury commented:
You can always reject the PR when it comes to your part of town.
It's a rare treat to see more than one in a round and I'm going to do what I can to add as much depth and texture as is appropriate with consistency and setting. It also gives more of a point to encryption keys. 

Currently pAI's can eavesdrop on their owner's headset, and as a being without much of a body they can spam common and pda's without worry. Also keep in mind that an EMP will put 'em out of commission for a good amount of time.

I see this PR as more of an opportunity to have individuals in departments who can engage the staff more socially and possibly coordinate some things to make everybody's life easier. The social aspect of this game is a big part of what sets it apart from others.

This is also a buff to pAI's translate ability. I frequently play a character with admin-disabled Galactic Common with a pAI translator. It's an absolute blast but it does get a bit difficult when my translator can't speak over the departmental channel to coordinate with my coworkers. I'm planning a "Never Learned Galactic Common" quirk and this ties into it pretty well.

Given the amount of shouting that is happening constantly and all the time at the station, I find it hard to believe that a pAI will powergame sec and command. Heck, once the AI and my Lawyer self witnessed someone plasmaflooding with the holodeck and despite both of our yelling in the sec channel, no one paid attention.

If you'd like to see pAI's become more than harm alarms maybe give them a chance to interact with their masters department and see what happens. A test merge isn't unreasonable.

Dorsisdwarf commented:
This PR body contains a great deal of writing and very little information about exactly what this PR does, especially on exactly what accesses this gives pAIs and how it works.

RaveRadbury commented:
> This PR body contains a great deal of writing and very little information about exactly what this PR does, especially on exactly what accesses this gives pAIs and how it works.

I've edited the body. Does it explain more now?

RaveRadbury committed:
Final version

It should all work.

# Conflicts:
#	code/game/objects/items/devices/radio/headset.dm
#	code/modules/mob/living/silicon/pai/pai.dm

RaveRadbury committed:
Merge

RaveRadbury committed:
Mergies

Dorsisdwarf commented:
👍

ExcessiveUseOfCobblestone commented on code/modules/mob/living/silicon/pai/pai.dm:
```diff
@@ -289,3 +291,13 @@
 /mob/living/silicon/pai/process()
 	emitterhealth = CLAMP((emitterhealth + emitterregen), -50, emittermaxhealth)
+
+/obj/item/paicard/attackby(obj/item/W, mob/user, params)
```
does this need to call inher?

ExcessiveUseOfCobblestone reviewed (commented):
Just want to make sure the attackby doesn't need inher (singals?) otherwise this is ok

ExcessiveUseOfCobblestone commented on code/modules/mob/living/silicon/pai/pai.dm:
```diff
@@ -289,3 +291,13 @@
 	emitterhealth = CLAMP((emitterhealth + emitterregen), -50, emittermaxhealth)
+
+/obj/item/paicard/attackby(obj/item/W, mob/user, params)
+	user.set_machine(src)
```
```suggestion
	..()
	user.set_machine(src)
```

ExcessiveUseOfCobblestone committed:
Update code/modules/mob/living/silicon/pai/pai.dm

Co-Authored-By: RaveRadbury <3204033+RaveRadbury@users.noreply.github.com>
